File: atlantic_report/atlantic_report/report/cust_payment_report___per_off_receipt/cust_payment_report___per_off_receipt.js
frappe.query_reports["Cust Payment Report - Per Off Receipt"] = {
    "filters": [
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            default: frappe.datetime.year_start(),
            reqd: 1
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            default: frappe.datetime.get_today(),
            reqd: 1
        }
    ],

    formatter: function (value, row, column, data, default_formatter) {

        value = default_formatter(value, row, column, data);

        // base style
        let style = "font-size: 12px; white-space: nowrap;";

        // ============================================================
        // STYLE UNTUK ROW TOTAL
        // ============================================================
        if (data && data.is_total_row) {
            style += `
                font-weight: bold;
                color: #000;
            `;
        }

        // ============================================================
        // ALIGNMENT PER KOLOM
        // ============================================================

        // Angka currency → right align
        if (["paid_amount", "paid_with"].includes(column.fieldname)) {
            return `<div style="${style} text-align:right;">${value}</div>`;
        }

        // Currency code (SGD, USD, etc.) → center
        if (["currency_amt", "currency_with"].includes(column.fieldname)) {
            return `<div style="${style} text-align:center;">${value}</div>`;
        }

        // Text columns → left align
        if (["pe_name", "date", "customer", "bank"].includes(column.fieldname)) {
            return `<div style="${style} text-align:left;">${value}</div>`;
        }

        return value;
    }
};

File: atlantic_report/atlantic_report/report/cust_payment_report___per_off_receipt/cust_payment_report___per_off_receipt.py
import frappe
from frappe.utils import flt

def execute(filters=None):
    if not filters:
        filters = frappe._dict()

    columns = get_columns()
    data = get_data(filters)
    return columns, data


# ================================================================
# COLUMNS
# ================================================================
def get_columns():
    return [
        {"label": "Off. Rec. No.", "fieldname": "pe_name", "fieldtype": "Data", "width": 140},
        {"label": "Date", "fieldname": "date", "fieldtype": "Date", "width": 100},
        {"label": "Customer", "fieldname": "customer", "fieldtype": "Data", "width": 240},

        {"label": "Ccy", "fieldname": "currency_amt", "fieldtype": "Data", "width": 50, "align": "center"},
        {"label": "Paid Amount", "fieldname": "paid_amount", "fieldtype": "Currency", "width": 130},

        {"label": "Ccy", "fieldname": "currency_with", "fieldtype": "Data", "width": 50, "align": "center"},
        {"label": "Paid With", "fieldname": "paid_with", "fieldtype": "Currency", "width": 130},

        {"label": "Bank", "fieldname": "bank", "fieldtype": "Data", "width": 180},
    ]


# ================================================================
# DATA QUERY
# ================================================================
def get_data(filters):

    sql = """
        SELECT
            pe.name AS pe_name,
            pe.posting_date AS date,
            pe.party AS customer,

            pe.paid_from_account_currency AS currency_amt,
            pe.paid_amount AS paid_amount,

            pe.paid_to_account_currency AS currency_with,
            pe.received_amount AS paid_with,

            acc.account_name AS bank_name

        FROM `tabPayment Entry` pe
        LEFT JOIN `tabAccount` acc ON acc.name = pe.paid_to

        WHERE pe.docstatus = 1
        AND pe.payment_type = 'Receive'
        AND pe.posting_date BETWEEN %(from_date)s AND %(to_date)s

        ORDER BY pe.name ASC
    """

    rows = frappe.db.sql(sql, filters, as_dict=True)
    result = []

    totals = {}  # contoh: {"SGD": {"paid_amount": X, "paid_with": Y}}

    for r in rows:

        # append detail rows
        result.append({
            "pe_name": r.pe_name,
            "date": r.date,
            "customer": r.customer,

            "currency_amt": r.currency_amt,
            "paid_amount": flt(r.paid_amount or 0),

            "currency_with": r.currency_with,
            "paid_with": flt(r.paid_with or 0),

            "bank": r.bank_name or ""
        })

        # determine currency grouping (use currency_amt as main currency)
        curr = r.currency_amt

        if curr not in totals:
            totals[curr] = {"paid_amount": 0, "paid_with": 0}

        totals[curr]["paid_amount"] += flt(r.paid_amount or 0)
        totals[curr]["paid_with"] += flt(r.paid_with or 0)


    # ================================================================
    # APPEND TOTAL ROWS BY CURRENCY
    # ================================================================
    for curr, t in totals.items():
        result.append({
            "pe_name": "",
            "date": "",
            "customer": f"<b>Total ({curr})</b>",

            "currency_amt": curr,
            "paid_amount": t["paid_amount"],

            "currency_with": curr,
            "paid_with": t["paid_with"],

            "bank": "",
            "is_total_row": 1
        })

    return result
